# 5. Resumen automático abstractivo

Resumen con un modelo ya entrenado (mT5_multilingual_XLSum) y con SLMs (zero-shot) sobre título + descripción + comentarios de cada hilo.

In [ ]:
from pathlib import Path

# Carpeta donde viven todos los .json del corpus (ver ../scripts/00_descarga_datos.py)
DATA_DIR = Path('../data')
assert DATA_DIR.exists(), "No se encuentra la carpeta data/ ¿ejecutaste desde notebooks/?"

## **5- Resumen automático abstractivo**

### 5.1- Usar un modelo ya entrenado para esta tarea, como mT5_multilingual_XLSum

In [ ]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
import torch

# Selección de dispositivo, línea de código que usabamos en prácticas para asignar gpu cuando sea posible
device = "cuda" if torch.cuda.is_available() else "cpu"

# Carga del modelo mT5 pre-entrenado
model_name_mt5 = "csebuetnlp/mT5_multilingual_XLSum"
tokenizer_mt5 = AutoTokenizer.from_pretrained(model_name_mt5)
model_mt5 = AutoModelForSeq2SeqLM.from_pretrained(model_name_mt5).to(device)

def resumen_mt5(titulo, descripcion, comentarios):
    texto_completo = f"Title: {titulo}. Description: {descripcion}. Comments: {' '.join(comentarios)}"

    if len(texto_completo) < 20: return "Texto demasiado corto."

    # Tokenización con truncado para ajustarse al límite del modelo
    inputs = tokenizer_mt5.encode("summarize: " + texto_completo,
                                  return_tensors="pt",
                                  truncation=True,
                                  max_length=512).to(device)

    with torch.no_grad():
        # Generación del resumen
        summary_ids = model_mt5.generate(inputs, max_length=150, min_length=30, length_penalty=2.0, num_beams=4)

    return tokenizer_mt5.decode(summary_ids[0], skip_special_tokens=True)

Loading weights:   0%|          | 0/284 [00:00<?, ?it/s]

[transformers] The tied weights mapping and config for this model specifies to tie shared.weight to lm_head.weight, but both are present in the checkpoints with different values, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning.


### 5.2- Usar SLMs (Small Large Models) como Gemma 3 (1B o 4B) o Llama 3.2 (1B)

In [ ]:
!pip install -U transformers accelerate bitsandbytes

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.5/10.5 MB 118.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.7/60.7 MB 14.8 MB/s eta 0:00:00
  Attempting uninstall: transformers
    Found existing installation: transformers 5.0.0
    Uninstalling transformers-5.0.0:
      Successfully uninstalled transformers-5.0.0


In [ ]:
# Importamos las librerias necesarias
from huggingface_hub import login

# Este token solo va a funcionar hasta una vez corregida la práctica, para poder ejecutar el cuaderno después de ese día teneis que incluir vuestro propio token.
login(token="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")

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
import torch

# Configuración de cuantización para reducir uso de memoria
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_quant_type="nf4"
)

# Definición de modelos SLM seleccionados
models = {
    "Llama-3.2": "meta-llama/Llama-3.2-1B-Instruct",
    "Gemma-3": "google/gemma-3-1b-it"
}

tokenizers = {}
loaded_models = {}

# Carga de modelos y tokenizadores
for name, path in models.items():
    print(f"Cargando {name}...")
    tokenizers[name] = AutoTokenizer.from_pretrained(path)
    loaded_models[name] = AutoModelForCausalLM.from_pretrained(
        path,
        quantization_config=bnb_config,
        device_map="auto"
    )

def generar_resumen_slm(model_name, subreddit, titulo, descripcion, comentarios):
    # Construcción del contexto del hilo con limitación de comentarios para control de longitud
    texto_hilo = f"Subreddit: {subreddit}\nTítulo: {titulo}\nDescripción: {descripcion}\nComentarios: {' '.join(comentarios[:10])}" # Limitamos comentarios por longitud
    # Prompt sencillo orientado a resumen
    prompt = f"""Eres un moderador experto del subreddit r/{subreddit}.
    Tu tarea es resumir el hilo de discusión que se presenta a continuación.

    Texto del hilo:
    {texto_hilo}

    Resumen:"""

    tokenizer = tokenizers[model_name]
    model = loaded_models[model_name]

    inputs = tokenizer(prompt, return_tensors="pt", truncation=True, max_length=2048).to(model.device)

    with torch.no_grad():
        outputs = model.generate(**inputs, max_new_tokens=80, temperature=0.1)

    # Post-procesamiento: Nos quedamos solo con lo generado después del prompt
    decoded = tokenizer.decode(outputs[0], skip_special_tokens=True)
    resumen = decoded.split("Resumen:")[-1].strip()
    return resumen

Cargando Llama-3.2...


Loading weights:   0%|          | 0/146 [00:00<?, ?it/s]

Cargando Gemma-3...


Loading weights:   0%|          | 0/340 [00:00<?, ?it/s]

**Justificación técnica de los modelos y límites**  
Para esta tarea, hemos comparado un modelo tradicional fine-tuneado con dos SLMs mediante aprendizaje sin ejemplos (Zero-shot Learning).  

**Modelo mT5 (CSEBUETNLP/mT5_multilingual_XLSum):**  
Este modelo cuenta con una limitación física de 512 tokens en su arquitectura de entrada.  
Al procesar hilos de Reddit que incluyen título, descripción y comentarios, superamos fácilmente este límite.  
Por ello, hemos aplicado una técnica de truncamiento, asegurando que el modelo reciba la información más relevante hasta llenar su memoria, ignorando el resto para evitar errores de procesamiento.  

**Modelos SLM (Llama-3.2-1B-Instruct y Gemma-3-1B-it):**  
Siguiendo las recomendaciones del enunciado, hemos seleccionado estos dos modelos para comparar sus capacidades de razonamiento.  
En nuestra implementación, hemos decidido incluir el título, la descripción y únicamente los primeros 10 comentarios de cada hilo.  
Esta decisión se tomó para mantener un equilibrio entre el contexto proporcionado y los límites de ventana de contexto de los modelos, evitando ruido excesivo o resúmenes demasiado diluidos.
Por otro lado, tal y como se indicaba en el enunciado, llevamos a cabo medidas de post-procesamiento, para quedarnos únicamente con el contenido generado a partir del marcador "Resumen:" que pusimos en nuestro prompt.

In [ ]:
import json
from pathlib import Path

# Configuración de rutas y subreddits
SUBREDDITS = ["Python", "Chemistry", "Investing", "Soccer", "Cooking", "Autism"]
BASE_PATH = Path('../data/')

for sub_name in SUBREDDITS:
    file_path = BASE_PATH / f"{sub_name}_filtrado.json"
    if not file_path.exists(): continue

    with open(file_path, 'r', encoding='utf-8') as f:
        data = json.load(f)

    print(f"Procesando r/{sub_name} (Título + Desc + Comentarios)...")

    for thread in data['submissions']:
        # Extracción de campos principales
        c_cuerpos = [c['body'] for c in thread.get('comments', [])]
        t = thread.get('title', '')
        d = thread.get('selftext', '')

        # 5.1 mT5 con todo el contenido (limitado a 512 tokens internamente)
        thread['summary_mT5'] = resumen_mt5(t, d, c_cuerpos)

        # 5.2 SLMs con todo el contenido
        thread['summary_Llama'] = generar_resumen_slm("Llama-3.2", sub_name, t, d, c_cuerpos)
        thread['summary_Gemma'] = generar_resumen_slm("Gemma-3", sub_name, t, d, c_cuerpos)

    # Guardamos el nuevo JSON
    output_path = BASE_PATH / f"{sub_name}_resumido_FINAL.json"
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

print("\n¡Hecho! Todos los hilos resumidos con el contenido completo.")

Procesando r/Python (Título + Desc + Comentarios)...


[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Ignoring clean_up_tokenization_spaces=True for BPE tokenizer TokenizersBackend. The clean_up_tokenization post-processing step is designed for WordPiece tokenizers and is destructive for BPE (it strips spaces before punctuation). Set clean_up_tokenization_spaces=False to suppress this warning, or set clean_up_tokenization_spaces_for_bpe_even_though_it_will_corrupt_output=True to force cleanup anyway.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_i

Procesando r/Chemistry (Título + Desc + Comentarios)...


[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_toke

Procesando r/Investing (Título + Desc + Comentarios)...


[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_toke

Procesando r/Soccer (Título + Desc + Comentarios)...


[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_toke

Procesando r/Cooking (Título + Desc + Comentarios)...


[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_toke

Procesando r/Autism (Título + Desc + Comentarios)...


[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_token_id` to `eos_token_id`:128001 for open-end generation.
[transformers] Setting `pad_toke


¡Hecho! Todos los hilos resumidos con el contenido completo.


**IMPORTANTE: No volver a ejecutar la siguiente celda, ya que las muestras de resúmenes fueron extraídas de forma aleatoria y no establecimos una semilla para reproducibilidad.**

**Entonces como el análisis cualitativo se basa en algunos ejemplos de resúmenes(de la salida obtenida), pues mejor que se puedan visualizar esos mismos ejemplos. Aún así, para asegurar, cuando nos refiramos a algún resumen en concreto, lo pondremos en el markdown por si acaso, para que quede constancia de a cual nos referimos.**

In [ ]:
import random
import json
from pathlib import Path

SUBREDDITS = ["Python", "Chemistry", "Investing", "Soccer", "Cooking", "Autism"]
BASE_PATH = Path('../data/')

def extraer_muestras_evaluacion(n=12):
    muestras = []
    for sub_name in SUBREDDITS:
        path = BASE_PATH / f"{sub_name}_resumido_FINAL.json"
        if path.exists():
            with open(path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                # Seleccionamos un par de hilos  de forma aleatoria de cada subreddit
                muestras.extend(random.sample(data['submissions'], 2))

    for i, m in enumerate(muestras[:n], 1):
        print(f"\n" + "="*80)
        print(f"HILO {i}: r/{m['subreddit']} - {m['title']}")
        print("-" * 80)
        print(f"mT5:   {m['summary_mT5']}")
        print(f"Llama: {m['summary_Llama']}")
        print(f"Gemma: {m['summary_Gemma']}")

extraer_muestras_evaluacion(12)


HILO 1: r/Python - Python 3.14 Released
--------------------------------------------------------------------------------
mT5:   This is a full transcript of the release of Python's version of 3.14. The release has been published by the BBC. Here is the full list of some of its significant changes, including improvements to the Python library.
Llama: Se presentan varias mejoras en el lenguaje de Python, incluyendo mejoras en el lenguaje de C, mejoras en el lenguaje de Python para la interfaz gráfica, y mejoras en el lenguaje de Python para la seguridad. Además, se presentan mejoras en el lenguaje de C, incluyendo mejoras en el
Gemma: Python 3.14 released.
    New features:
    * PEP 649 and PEP 749: Deferred evaluation of annotations
    * PEP 734: Multiple interpreters in the standard library
    * PEP 750: Template strings
    * PEP 758: Allow except and except* expressions without brackets
    *

HILO 2: r/Python - Python Foundation goes ride or DEI, rejects government grant with st

**Metodología de evaluación**  
Para realizar la evaluación cualitativa, hemos seleccionado una muestra de 12 hilos(aunque en el enunciado se mencionan 10, de esta forma nos podemos quedar con un par por subreddit) de forma aleatoria, repartidos equitativamente (2 por subreddit).  
Esto nos permite observar cómo se comportan los modelos ante temáticas tan diversas como la programación (r/Python), la ciencia (r/chemistry) o el deporte (r/soccer).

**Análisis Cualitativo**  
Tras analizar las salidas, nuestra conclusión general es que, aunque los modelos mantienen la temática del hilo, la calidad de los resúmenes deja mucho que desear si los comparamos con LLMs de mayor escala.  

**Alucinaciones en mT5:**  
Hemos detectado un patrón de error grave.  
Debido a que mT5 fue entrenado con el dataset de noticias XLSum, tiende a forzar el formato de "noticia de la BBC".  
Un ejemplo claro es el Hilo 12 (r/autism), donde mT5 afirma que el texto es un "transcript de la serie The Boss de la BBC sobre Michael Jackson", cuando en realidad es un usuario contando un incidente en una tienda.

Resumen referido: "HILO 12: r/autism - Got apprehended for leaving a store without buying anything

mT5:   The BBC's weekly The Boss series profiles different people from around the world. This is a full transcript of how Michael Jackson started the "mask trend" and why he was stopped from stealing Lego sets. Here, she explains what happened to her and how she feels after leaving the B&M store without buying anything.
"

**Comportamiento de Llama-3.2:**  
Hemos observado una fuerte tendencia de este modelo a responder en español, a pesar de que el contenido original de los hilos está en inglés.  
Esto se debe probablemente a su fuerte alineamiento multilingüe y a que el prompt de instrucciones se le pasó en español.  
Aunque el resumen es coherente, a menudo se corta bruscamente por el límite de tokens de salida.  

Ejemplo de resumen: "HILO 10: r/Cooking - Help me settle an argument with my wife.

Llama: La respuesta de la persona que se refiere a sí misma es que blanching en agua caliente no es necesariamente una forma de limpiar los vegetales, sino más bien una forma de prepararlos. La persona que se refiere a sí misma también menciona que no hay diferencias culturales en la limpieza de los vegetales, y que la persona que se"

**Comportamiento de Gemma-3:**  
A diferencia de Llama, Gemma suele mantenerse en inglés y prefiere estructuras de listas(como se ve en el Hilo 1 o el Hilo 5), algo que facilita la lectura.

Resúmenes referidos: "HILO 1: r/Python - Python 3.14 Released

Gemma: Python 3.14 released.
    New features:
    * PEP 649 and PEP 749: Deferred evaluation of annotations
    * PEP 734: Multiple interpreters in the standard library
    * PEP 750: Template strings
    * PEP 758: Allow except and except* expressions without brackets
    *"


"HILO 5: r/investing - Trump says company should only report earnings every 6 months

Gemma: Trump says company should only report earnings every 6 months.
    The US has done fine reporting quarterly.
    The US needs to shift to a more transparent reporting cadence.
    Retail investors should prioritize long-term deliverables over quick wins.
    The Great Depression was caused by a lack of transparency.
    The current quarterly reporting system is inefficient and creates more pressure."

**Pequeño inciso:**

Como pequeño "error" o posible mejora, no caímos antes de extraer los resúmenes, tal y como mencionaste en las clases de teoría, en poner el prompt en inglés, ya que sobre todo las SLMs, y realmente casi todos los modelos de lenguaje, están entrenados mayoritariamente con texto en inglés, entonces seguramente con un prompt en inglés hubiesemes obtenido ligeras mejoras, como por lo menos la salida de Llama en inglés, en vez de español. Esto no es ninguna excusa para justificarnos de no haberlo hecho de esa forma; simplemente lo llevamos a cabo con el prompt en español, primeramente porque en los ejemplos de las sesiones de prácticas aparecían ejemplos en español, y como hemos mencionado, porque no caímos. Pero lo mencionamos para que quede constancia de ello, ya que este apartado fue realizado en varias ocasiones, haciéndolo de distintas formas, y esperando repetidas veces tiempos largos de cómputo, entonces como consideramos que los resultados obtenidos y el análisis realizado están considerablemente bien, pues no vamos a realizarlo de nuevo, porque reiteramos en que al final es prácticamente lo mismo y los cambios no serían tan grandes, dando lugar a una salida parecida y un mismo análisis.